rclcpp/src/rclcpp/executors/detail/timer_manager.hpp

```diff
@@ -0,0 +1,356 @@
+  }
+
+  void call_ready_timer_callbacks()
+  {
+    auto readd_timer_to_running_map = [this](TimerMap::node_type && e)
+      {
+        const auto & timer_data = e.mapped();
+        if(remove_if_dropped(timer_data))
+        {
+          return;
+        }
+
+        int64_t next_call_time;
+
+        auto ret = rcl_timer_get_next_call_time(timer_data->rcl_ref.get(), &next_call_time);
+
+        if (ret == RCL_RET_OK) {
+          e.key() = std::chrono::nanoseconds(next_call_time);
+          running_timers.insert(std::move(e));
+        }
+      };
+
+    while (!running_timers.empty()) {
+
+      if(remove_if_dropped(running_timers.begin()->second))
+      {
+        continue;
+      }
+
+      int64_t time_until_call;
+
+      const rcl_timer_t * rcl_timer_ref = running_timers.begin()->second->rcl_ref.get();
+      auto ret = rcl_timer_get_time_until_next_call(rcl_timer_ref, &time_until_call);
+      if (ret == RCL_RET_TIMER_CANCELED) {
+        running_timers.erase(running_timers.begin());
+        continue;
+      }
+
+      if (time_until_call <= 0) {
+        // advance next call time;
+        rcl_ret_t ret = rcl_timer_call(const_cast<rcl_timer_t *>(rcl_timer_ref));
+        if (ret == RCL_RET_TIMER_CANCELED) {
+          running_timers.erase(running_timers.begin());
+          continue;
+        }
+
+        // timer is ready, execute callback
+        running_timers.begin()->second->timer_ready_callback();
+        readd_timer_to_running_map(running_timers.extract(running_timers.begin()));
+        continue;
+      }
+      break;
+    }
+  }
+
+  void timer_thread()
+  {
+    while (running && rclcpp::ok()) {
+      std::chrono::nanoseconds next_wakeup_time;
+      {
+        std::scoped_lock l(mutex);
+        call_ready_timer_callbacks();
+
+        next_wakeup_time = get_next_timer_ready_time();
+      }
+      try {
+//             RCUTILS_LOG_ERROR_NAMED("rclcpp", "TimerQueue::timer_thread before sleep, next wakeup time %+" PRId64 , next_wakeup_time.count());
+        used_clock_for_timers.sleep_until(rclcpp::Time(next_wakeup_time.count(), timer_type));
+      } catch (const std::runtime_error &) {
+        //there is a race on shutdown, were the context may become invalid, while we call sleep_until
+        running = false;
+      }
+    }
+    thread_terminated = true;
+  }
+
+  rcl_clock_type_t timer_type;
+
+  Context::SharedPtr clock_sleep_context;
+
+  rclcpp::Clock used_clock_for_timers;
+
+  std::mutex mutex;
+
+  std::atomic_bool running = true;
+  std::atomic_bool thread_terminated = false;
+
+  std::vector<std::unique_ptr<TimerData>> all_timers;
+
+  using TimerMap = std::multimap<std::chrono::nanoseconds, const TimerData *>;
+  TimerMap running_timers;
+
+  std::thread trigger_thread;
+
+  std::condition_variable thread_conditional;
+};
+
+class TimerManager
+{
+  std::array<TimerQueue, 3> timer_queues;
+
+public:
+  TimerManager()
+  : timer_queues{RCL_ROS_TIME, RCL_SYSTEM_TIME, RCL_STEADY_TIME}
```

Comment:
queues should be created when someone tries to add the first timer of that type, and destroyed when empty

Reply:
Hm, I don't see any harm in having two idle threads in this case. They would block on the conditional and consume no CPU whatsoever. 

Starting threads on runtime, and additional checks inside the executor code I see more as a problem though.

Reply:
Even if there will be no CPU consumption, threads also have a memory impact.
Moreover, even if the impact is likely very small, most systems will only have 1 type of timers (a lot of users use the default RCL_ROS_TIME clock), it seems strange that if you have 10 executors with 1 timer each you will create 40 threads instead of 20.

I agree that stopping and restarting them is probably not great, but at least we should wait to create a thread only if we need it.
This is an issue also with the current implementation, that always creates 1 thread even if you have no timers... But it's worst here because we always create 3 threads.

Reply:
Still don't see the issue, memory impact is at most 100 KB per thread. So event for a really bit system like ours (106 executors), we are talking 30 MB of memory vs 10MB.

Reply:
In embedded applications, knowing that there's 2-3 MB wasted on threads that are never used is a problem, and something that can be easily optimized.

I agree that starting and stopping the threads is likely not the correct solution (at least it shouldn't be the default behavior), but I don't see any downside in delaying the creation of the timer queues until they are needed.

The `remove_timer` and `add_timer` functions below should just look up the timer clock type and create a queue if they don't have one already, this also makes the code simpler because we don't have to "try" to add the timer to all queue, but we already know in which queue it should be added and we can check a return code for success/failures.